### DISCUSSION TITLE
Share my findings on Wyze cam v3 upgrade firmware image and flash chip

### DISCUSSION THREAD

**Discussion by archandanime** [gtxaspec/wz_mini_hacks#633] [Development]
There are people who already knew this, but I want to share my findings on Wyce Cam V3, hope it helps with future developments of wz_mini_hacks!
It would be too long if I put on how I figured out these things, so I'm only putting what worked here.

### Disclaimer: If you attempt to do this, backup all your partitions first., or at least `/configs`, typing a wrong character can make the Camera hard-bricked (not too serious but it can happen).

I have Wyze Cam V3 `demo_wcv3.bin` (4.36.9.139)
Analyze it:
```bash
> binwalk demo_wcv3.bin_wcv3_4.36.9.139

DECIMAL       HEXADECIMAL     DESCRIPTION
--------------------------------------------------------------------------------
0             0x0             uImage header, header size: 64 bytes, header CRC: 0xCC3249BD, created: 2022-05-13 20:22:40, image size: 9912320 bytes, Data Address: 0x0, Entry Point: 0x0, data CRC: 0x5A46572, OS: Linux, CPU: MIPS, image type: Firmware Image, compression type: none, image name: "jz_fw"
64            0x40            uImage header, header size: 64 bytes, header CRC: 0x3CEC2718, created: 2022-03-02 08:42:59, image size: 1897330 bytes, Data Address: 0x80010000, Entry Point: 0x80416900, data CRC: 0x1589A43F, OS: Linux, CPU: MIPS, image type: OS Kernel Image, compression type: lzma, image name: "Linux-3.10.14__isvp_swan_1.0__"
128           0x80            LZMA compressed data, properties: 0x5D, dictionary size: 67108864 bytes, uncompressed size: -1 bytes
2031680       0x1F0040        Squashfs filesystem, little endian, version 4.0, compression:xz, size: 3853780 bytes, 384 inodes, blocksize: 131072 bytes, created: 2022-05-13 20:22:38
6029376       0x5C0040        Squashfs filesystem, little endian, version 4.0, compression:xz, size: 3879930 bytes, 196 inodes, blocksize: 131072 bytes, created: 2022-05-13 20:22:39
```
So it contains a kernel image and 2 squashfs images. Let's extract it!
```bash
> dd if=demo_wcv3.bin_wcv3_4.36.9.139 of=demo_wcv3.bin_wcv3_4.36.9.139_kernel bs=1 count=2031680
2031680+0 records in
2031680+0 records out
2031680 bytes (2.0 MB, 1.9 MiB) copied, 4.73143 s, 429 kB/s
> echo $(( 6029376 - 2031680 ))
3997696
> dd if=demo_wcv3.bin_wcv3_4.36.9.139 of=demo_wcv3.bin_wcv3_4.36.9.139_squashfs_1 bs=1 count=3997696 skip=2031680
3997696+0 records in
3997696+0 records out
3997696 bytes (4.0 MB, 3.8 MiB) copied, 9.23128 s, 433 kB/s
> dd if=demo_wcv3.bin_wcv3_4.36.9.139 of=demo_wcv3.bin_wcv3_4.36.9.139_squashfs_2 bs=1 skip=6029376
3883072+0 records in
3883072+0 records out
3883072 bytes (3.9 MB, 3.7 MiB) copied, 8.83351 s, 440 kB/s
```

Now figure out what's inside these 2 squashfs images:
```bash
> mkdir squash_1 squashfs_2
> sudo mount demo_wcv3.bin_wcv3_4.36.9.139_squashfs_1 squash_1
> sudo mount demo_wcv3.bin_wcv3_4.36.9.139_squashfs_2 squash_2
> ls squash_1
aback  configs  etc    lib      media  opt   root  sbin  system    tmp  var
bin    dev      kback  linuxrc  mnt    proc  run   sys   thirdlib  usr
> ls squash_2
Upgrade  bin  driver  etc  init  lib  local
```
This is familiar, now we know that these 2 squashfs files are actually `rootfs` (mounted at `/`)and `driver`(mounted at `/system`) partitions

Then we rename them to `kernel.img`, `rootfs.img` and `driver.img`:
```bash
> file kernel.img driver.img rootfs.img
kernel.img: u-boot legacy uImage, jz_fw, Linux/MIPS, Firmware Image (Not compressed), 9912320 bytes, Fri May 13 20:22:40 2022, Load Address: 00000000, Entry Point: 00000000, Header CRC: 0XCC3249BD, Data CRC: 0X5A46572
driver.img: Squashfs filesystem, little endian, version 4.0, xz compressed, 3879930 bytes, 196 inodes, blocksize: 131072 bytes, created: Fri May 13 20:22:39 2022
rootfs.img: Squashfs filesystem, little endian, version 4.0, xz compressed, 3853780 bytes, 384 inodes, blocksize: 131072 bytes, created: Fri May 13 20:22:38 2022
```

Let try to modify  `rootfs.img`, by extracting, modifying then repacking:
```bash
> unsquashfs rootfs.img
> ls squashfs-root
aback  configs  etc    lib      media  opt   root  sbin  system    tmp  var
bin    dev      kback  linuxrc  mnt    proc  run   sys   thirdlib  usr
> touch squashfs-root/this_has_been_modified
> sudo chown root:root -R squashfs-root/
> mksquashfs squashfs-root rootfs_modified.img -comp xz
```

We have obtained a modified `rootfs` mage, let's try writing it to Camera flash chip:
```bash
> scp -O rootfs_modified.img root@192.168.10.20:/opt
rootfs_modified.img                                                                 100% 3764KB   1.4MB/s   00:02    
```

```
[root@WCV3-Indoor:opt]# cat /proc/mtd 
dev:    size   erasesize  name
mtd0: 00040000 00008000 "boot"
mtd1: 001f0000 00008000 "kernel"
mtd2: 003d0000 00008000 "rootfs"
mtd3: 003d0000 00008000 "app"
mtd4: 001f0000 00008000 "kback"
mtd5: 003d0000 00008000 "aback"
mtd6: 00060000 00008000 "cfg"
mtd7: 00010000 00008000 "para"
[root@WCV3-Indoor:opt]# flashcp rootfs_modified.img /dev/mtd2
[root@WCV3-Indoor:opt]# reboot
[root@WCV3-Indoor:opt]# Connection to 192.168.10.20 closed by remote host.
Connection to 192.168.10.20 closed.
```

Finally check the new rootfs:
```bash
> ssh root@192.168.10.20
COLUMNS=118;LINES=55;export COLUMNS LINES;
[root@WCV3-Indoor:~]# ls /
aback                   lib                     root                    this_has_been_modified
bin                     linuxrc                 run                     tmp
configs                 media                   sbin                    usr
dev                     mnt                     sys                     var
etc                     opt                     system
```
We have successfully modified the rootfs, this change is permanent unless you restore backed up partition.

**Comment by archandanime** [comment]
So what we have learned:
- Wyze cam v3 upgrade firmware image contains only `kernel`, `rootfs` and `driver`(also called `system` or `app`).
- `rootfs` and `driver` partitions are in squashfs format using `xz` compression. If you try to repack the image after modification, use `-comp xz` because it the only compression type that the kernel supports, `mksquashfs` uses `gzip` by default.
- "Partitions" inside upgrade firmware files are smaller than the actually partitions on the flash chip, we can add contents to them as long as new partition file is smaller than physical partition on the flash chip.

**Comment by archandanime** [comment]
I'm still figuring why the camera doesn't flash my custom image from:
```
> cat kernel.img rootfs_modified.img driver.img > demo_wcv3.bin
```

The reason is the uboot has some sort of verification:
```
Filesystem: FAT32 "INDOOR-SD  "
file_fat_detectfs OK
reading factory_t31_ZMC6tiIDQN
factory_t31_ZMC6tiIDQN not found
gpio_request lable = sdupgrade gpio = 51
the manufacturer 1c
SF: Detected EN25QH128A

Interface:  MMC
  Device 0: Vendor: Man 0000fe Snr 000caa01 Rev: 0.0 Prod: SD16G 
            Type: Removable Hard Disk
            Capacity: 29818.0 MB = 29.1 GB (61067264 x 512)
Filesystem: FAT32 "INDOOR-SD  "
file_fat_detectfs OK
reading demo_wcv3.bin
reading demo_wcv3.bin
demo_wcv3.bin sigh verify fail
the manufacturer 1c
SF: Detected EN25QH128A
```

**Reply by gtxaspec** [reply]
It validates some sort of magic bytes in the image, I don't believe it's signed though.   

Theres a special repacker tool to repack the image as well, check the DaFangHacks repo, you'll need it to create a valid bin anyway.

@HclX had done some research on the demo.bin for the v3, I would check his repos: https://github.com/HclX/WyzeHacks

Alternatives:

boot from kernel + custom initramfs
- have the initramfs mount the sd card and flash your file with a custom script reading everything off the sd card

(risky) write an invalid u-boot to /dev/mtd0.  use the "cloner" ingenic tool to flash your partitions manually over usb.  If the t31 can not load u-boot, it proceeds to enable a special USB interface that you can use to recover the flash via USB + PC tools 

make sure you backup /configs or /params of course.

These cameras are pretty unbrickable.

**Reply by archandanime** [reply]
Thanks for the recommendation, I'll check it out.

> (risky) write an invalid u-boot to /dev/mtd0. use the "cloner" ingenic tool to flash your partitions manually over usb. If the t31 can not load u-boot, it proceeds to enable a special USB interface that you can use to recover the flash via USB + PC tools

Wow, this sounds like some Mediatek chips, I had a Motorola phone with that chip and it was impossible to brick. Thanks for telling me, this would help a lot(but it's the last resort).

Booting from kernel + custom initramfs works, but if I want to build a wz_mini_hack without SD Card(and share to people), letting uboot flash the image seems to be a safe option.

**Reply by archandanime** [reply]
> V3 camera introduced signature verification on demo.bin. Unless we know Wyze's private signing key, or we found a vulnerabilty in the related code, a modified demo.bin won't be accepted.
> Luckily, the signature verification is only for SD card recovery, not for booting. So once you find a way to modify the flash content, the bootloader will happily load whatever in the flash. That leads to this method.

[source](https://github.com/HclX/WyzeHacks/blob/0ca2fc03af156a364165d1a082c8c6fbfc2e1fb9/tools/v3/README.md)

But somehow, [someone](https://github.com/CedBri/wyzecamv3) figured out the way(?). I tried but it didn't work.

**Comment by archandanime** [comment]
I wonder if someone can get a custom `demo_wcv3.bin` work with [this method](https://github.com/CedBri/wyzecamv3):
I did but it failed:
```bash
> file demo_wcv3.bin
demo_wcv3.bin: u-boot legacy uImage, jz_fw, Linux/MIPS, Firmware Image (Not compressed), 9912320 bytes, Fri May 13 20:22:40 2022, Load Address: 00000000, Entry Point: 00000000, Header CRC: 0XCC3249BD, Data CRC: 0X5A46572
> ./wyze_extractor.py unpack demo_wcv3.bin 
> sudo unsquashfs squashfs_1
> sudo nano  squashfs-root/etc/shadow
> sudo nano squashfs-root/etc/init.d/rcS 
> mv squashfs_1 squashfs_1_old
> sudo  mksquashfs squashfs-root/ squashfs_1 -comp xz -b 131072
> mkimage -A MIPS -O linux -T firmware -C none -a 0 -e 0 -n jz_fw -d exploited_wyze.bin demo_wcv3_exploited.bin
> file demo_wcv3_exploited.bin
demo_wcv3_exploited.bin: u-boot legacy uImage, jz_fw, Linux/MIPS, Firmware Image (Not compressed), 9883648 bytes, Fri Oct  6 15:40:19 2023, Load Address: 00000000, Entry Point: 00000000, Header CRC: 0X5BAB538E, Data CRC: 0XC38A8FF0
```

I used `demo_wcv3_exploited.bin` from the above repo and it didn't work either :))
```
file_fat_detectfs OK
reading demo_wcv3.bin
reading demo_wcv3.bin
demo_wcv3.bin sigh verify fail
```

**Comment by archandanime** [comment]
Wyze did put uboot source code here: https://support.wyze.com/hc/en-us/articles/360012546832-Open-Source-Software, let me see if there's something interesting. Maybe there's a way to bypass the verification.

**Comment by archandanime** [comment]
So there isn't a way to flash a custom `demo_wcv3.bin` because the file is signed:
./common/cmd_sdupdate.c
```c
#ifdef CONFIG_AUTO_UPDATE_CRYPTO
        if(0 != eddsa_verify(LOAD_ADDR, sz))
        {
                printf("%s sigh verify fail\n", aufile[idx]);
                return -1;
        }
        printf("%s sigh verify success\n", aufile[idx]);
        sz -= 64;
#endif
```
even though building a custom uboot is possbile with the source code as OpenIPC already did. Maybe because my device was produced(2021) after Wyze made changes to the firmware?

**Comment by archandanime** [comment]
~~Ingenic T31 boot sequence + debricking with SD Card: https://github.com/OpenIPC/wiki/blob/master/en/unbrick-ingenic-t31-with-sd-card.md~~ (Doesn't work with Wyze Cams)

>~~If failed to boot from uboot on flash memory, T31 will then try to boot from SD card, no matter what bootsel pin is set. So if the uboot on flash chip is somehow broken, we can burn uboot to a SD card and boot from it. The uboot file should be specially compiled for SD card boot, you can't use the one for normal flash boot.~~

**Reply by gtxaspec** [reply]
Doesn't work on haulai devices, they use a transistor controlled by gpio to enable or disable the memory card voltage.

**Comment by archandanime** [comment]
~~If you try to modify `rootfs` partition, make sure it's the same version as `app` partiton, or else the device would have boot loop.~~
```
===========welcome to ver-comp tool=========
[ver-comp]dbg: appver:  4.36.9.139
[ver-comp]dbg: rootver: 4.36.11.5859
```
Edit: It's wrong. Newer firmware on this camera is really unusable, even from a clean upgrade.

**Comment by archandanime** [comment]
By modifying `(rootfs partition)/usr/app.ver` and `(app partition)/bin/app.ver`, Wyze app shows that my firmware is up-to-date:
```
[root@WyzeCamv3:~]# cat /usr/app.ver /system/bin/app.ver 
[VER]
appver=9.36.3.19
[VER]
appver=9.36.9.139
```
![Screenshot_20231012-211439_Wyze](https://github.com/gtxaspec/wz_mini_hacks/assets/36922812/469e92e1-af94-4854-a5ec-275a57c438d8)

**Comment by archandanime** [comment]
When upgrading to lastest firmware version, I saw this in a file on `/tmp/`:
```bash
#!/bin/sh

if [ -f /tmp/Upgrade/SpotlightUpgradeRun.sh ]; then
	echo "to upgrade spotlight..."
	/tmp/Upgrade/SpotlightUpgradeRun.sh
else
	echo "no use upgrade spotlight..."
fi

if [ -f /tmp/Upgrade/config_repair_up.sh ]; then
	sh /tmp/Upgrade/config_repair_up.sh
else
	echo "not found config_repair_up..."
fi

echo "dump existing rootfs !!!!!!!!!!!"
dd if=/dev/mtd2 of=/tmp/Upgrade/rootfs.back

#backup current rootfs so we can recover it if we lost power
echo "erase backa !!!!!!!!!!!"
flash_eraseall /dev/mtd5
sync
echo "write backa for backup rootfs !!!!!!!!!!!"
flashcp -v /tmp/Upgrade/rootfs.back /dev/mtd5
sync

echo "erase para !!!!!!!!!!!"
flash_eraseall /dev/mtd7
sync
echo "write para !!!!!!!!!!!"
flashcp -v /tmp/Upgrade/PARA.rootfs /dev/mtd7
sync

#now write the new rootfs directly
echo "erase rootfs !!!!!!!!!!!"
flash_eraseall /dev/mtd2
sync
echo "write rootfs !!!!!!!!!!!"
flashcp -v /tmp/Upgrade/rootfs /dev/mtd2
sync

#clear the rootfs flag
echo "erase para !!!!!!!!!!!"
flash_eraseall /dev/mtd7
sync
echo "write para !!!!!!!!!!!"
flashcp -v /tmp/Upgrade/PARA.null /dev/mtd7
sync

#write app to backa
echo "erase backa !!!!!!!!!!!"
flash_eraseall /dev/mtd5
sync
echo "write backa !!!!!!!!!!!"
flashcp -v /tmp/Upgrade/app /dev/mtd5
sync

echo "erase para !!!!!!!!!!!"
flash_eraseall /dev/mtd7
sync
echo "write para !!!!!!!!!!!"
flashcp -v /tmp/Upgrade/PARA /dev/mtd7
sync
sync

```
So the `aback` - `/dev/mtd5` actually has a purpose: It is used to backup current `rootfs` and `app` partitions in case of power lost. After current `rootfs` and `app` partitions are written into `aback`, partition `para` - `/dev/mtd7` is updated.
I'm not sure but I speculate that `para` partition content is checked by the stock kernel with a flag: `rootfs` or `app` or `null`. If it's not `null`, backup content of the `rootfs` or `app` partition is written back to its corresponding partition for recovery.

**Comment by archandanime** [comment]
I also made a tool to backup/restore partitions to make flashing custom partition less hassling: [here](https://github.com/archandanime/wcv3_flash-helper). It only works on Wyze Cam v3 though.

**Comment by archandanime** [comment]
I have found out that Wyze has updated rootfs partition only once by downloading all possible firmware file using URL from [WyzeUpdater](https://github.com/HclX/WyzeUpdater).
```bash
> for i in *.tar; do echo -n $(basename -s .tar $i); tar -axf $i Upgrade/PARA -O | sed 's/FWGRADEUP=/\ \ /g';done
4.36.0.125  kernel+app
4.36.0.228  kernel+app
4.36.0.248  kernel+app
4.36.0.252  kernel+app
4.36.0.280  app
4.36.1.4  app
4.36.10.2700  app
4.36.10.2864  app
4.36.10.3406  app
4.36.10.4054  app
4.36.11.4679  app
4.36.11.5859  app
4.36.2.5  kernel+app
4.36.3.18  kernel+app
4.36.3.19  kernel+rootfs
4.36.6.17  kernel+app
4.36.7.22  kernel+app
4.36.8.15  kernel+app
4.36.8.26  kernel+app
4.36.8.32  kernel+app
4.36.9.131  kernel+app
4.36.9.139  kernel+app
```
From their upgrade `upgraderun.sh` script:
```bash
echo "erase backk !!!!!!!!!!!"
flash_eraseall /dev/mtd4
sync
echo "write backk !!!!!!!!!!!"
flashcp -v /tmp/Upgrade/kernel /dev/mtd4
sync

echo "erase backa !!!!!!!!!!!"
flash_eraseall /dev/mtd5
sync
echo "write backa !!!!!!!!!!!"
flashcp -v /tmp/Upgrade/app /dev/mtd5
sync
```
`backa`(or `aback`) actually stands for "app backup" and `backk`(or `kback`) stands for "kernel backup".
`para` partition is used by the bootloader to trigger copy from `aback` to `app` partition and `kback` to `kernel` partition.
Example `para` partition content:
```
FWGRADEUP=kernel+app
```

From U-Boot [source](https://support.wyze.com/hc/en-us/articles/360012546832-Open-Source-Software):
```bash
> grep -r FWGRADEUP  .
./common/cmd_sdupdate.c:			if(strncmp(buf+j, "FWGRADEUP=", 10) == 0)
./common/cmd_sdupdate.c:			printf("FWGRADEUP=kernel+app!\n");
./common/cmd_sdupdate.c:			printf("FWGRADEUP=kernel+rootfs!\n");
./common/cmd_sdupdate.c:			printf("FWGRADEUP=Kernel!\n");
./common/cmd_sdupdate.c:			printf("FWGRADEUP=rootfs!\n");
./common/cmd_sdupdate.c:			printf("FWGRADEUP=app!\n");
...
> nano ./common/cmd_sdupdate.c
                    memcpy(pbuf+ausize[IDX_KBACK], ppbuf, ausize[IDX_RABACK]);
```

`backa` is basically the staging partition for the bootloader to update `app` or `rootfs`, they can not be updated at the same time.

**Comment by archandanime** [comment]
With firmware upgrade version `4.36.3.19`. I made a comparision between rootfs before upgrade `4.36.6.17` and `4.36.0.125`
[diff.txt](https://github.com/gtxaspec/wz_mini_hacks/files/12903894/diff.txt)

```ash
> unsquashfs 4.36.0.125.sqsh && mv squashfs-root 4.36.0.125
> unsquashfs 4.36.6.17.sqsh && mv squashfs-root 4.36.6.17
> find 4.36.0.125 -type l -delete
> find 4.36.6.17 -type l -delete
> for f in `find 4.36.6.17/* ! -type l`; do diff -rq $f 4.36.0.125/${f##*/} 2> /dev/null  ; done
Files 4.36.6.17/bin/busybox and 4.36.0.125/bin/busybox differ
Only in 4.36.6.17/bin: cttyhack
Only in 4.36.6.17/bin: iwconfig
Only in 4.36.6.17/bin: iwgetid
Only in 4.36.6.17/bin: iwlist
Only in 4.36.6.17/bin: iwpriv
Only in 4.36.6.17/bin: uname
Files 4.36.6.17/bin/wpa_supplicant and 4.36.0.125/bin/wpa_supplicant differ
Files 4.36.6.17/etc/shadow and 4.36.0.125/etc/shadow differ
Files 4.36.6.17/thirdlib/libcrypto.so.1.1 and 4.36.0.125/thirdlib/libcrypto.so.1.1 differ
Only in 4.36.6.17/thirdlib: libcurl.so.4.7.0
Only in 4.36.6.17/thirdlib: libnl-3.so.200.21.0
Only in 4.36.6.17/thirdlib: libnl-genl-3.so.200.21.0
Files 4.36.6.17/thirdlib/libssl.so.1.1 and 4.36.0.125/thirdlib/libssl.so.1.1 differ
Files 4.36.6.17/usr/app.ver and 4.36.0.125/usr/app.ver differ
Only in 4.36.0.125/bin: busybox
Only in 4.36.0.125/bin: hostapd
Only in 4.36.6.17/usr/bin: iperf
Only in 4.36.0.125/bin: wpa_cli
Only in 4.36.0.125/bin: wpa_supplicant
Only in 4.36.0.125/sbin: flash_erase
```
There is no significant change beside changing `/etc/shadow`. I believe they don't have a reason to update `rootfs` again in the future.
